More getters for VariantContext

magicDGS opened the pull request:
### Description
- Fixes #371 including list getters as String, Integer and Double.
### Checklist
- [x] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

magicDGS commented:
@yfarjoun and @ldgauthier, because you are the ones interested in that issue, could you have a look here?

yfarjoun commented on src/main/java/htsjdk/variant/variantcontext/CommonInfo.java:
```diff
@@ -26,6 +26,7 @@
 package htsjdk.variant.variantcontext;
 
 
+
```
more space?

yfarjoun commented on src/main/java/htsjdk/variant/variantcontext/CommonInfo.java:
```diff
@@ -255,6 +258,40 @@ public Object getAttribute(String key, Object defaultValue) {
+                            if (x == null || x == VCFConstants.MISSING_VALUE_v4) {
+                                return defaultValue;
+                            } else if (x instanceof Integer) {
+                                return (Integer) x;
```
what if someone writes an array of ints? int[5] to the attribute? will it work?

magicDGS replied:
`int[5]` generates an empty array with each element initialized to 0, so it will work. The same for the double.

yfarjoun commented on src/main/java/htsjdk/variant/variantcontext/CommonInfo.java:
```diff
@@ -255,6 +258,40 @@ public Object getAttribute(String key, Object defaultValue) {
+                    } else if (x instanceof Double) {
+                        return (Double) x;
+                    } else {
+                        return Double.valueOf((String)x); // throws an exception if this isn't a string
```
same question about double[]

magicDGS commented:
@yfarjoun, before adding the tests. Do you think that it is better to have as a default value a `Supplier<Integer>` instead of an `int`. This will allow to generate `null` values in the list if the user want's to now if the value was filled with the default or it have teh value...

magicDGS commented:
@yfarjoun I added the tests and I realized that I based my implementation on (`public List<Object> getAttributeAsList(String key) `) was broken for primitive type arrays. Thus I added a simple patch checking for double/int arrays, and boxing the values with stream processing.

yfarjoun commented on src/main/java/htsjdk/variant/variantcontext/CommonInfo.java:
```diff
@@ -251,10 +254,51 @@ public Object getAttribute(String key, Object defaultValue) {
+
+    public List<Integer> getAttributeAsIntList(String key, Integer defaultValue) {
+        return getAttributeAsList(key, x -> {
+                            if (x == null || x == VCFConstants.MISSING_VALUE_v4) {
```
pull out one or two indents.

magicDGS replied:
Done

yfarjoun commented:
looks good. thanks @magicDGS. minor comments (and looks like you need to rebase) and we should be ready to go!

magicDGS commented:
Rebased and addressed comments. Back to you @yfarjoun!

yfarjoun commented on src/test/java/htsjdk/variant/variantcontext/VariantContextUnitTest.java:
```diff
@@ -1478,4 +1477,71 @@ public void testExtractStructuralVariationsData(final File vcfFile) {
+        Assert.assertEquals(context.getAttributeAsDoubleList("StringList", 5), Arrays.asList(1d, 2d));
+        try {
+            context.getAttributeAsIntList("NotNumeric", 5);
+            Assert.fail();
```
what's the point of this try/catch block?

magicDGS replied:
This assers that an exception is thrown if a numeric list is requested for a key that is not a number. I change it here and the rest of tests for `Assert.assertThrows`, which is clear.

yfarjoun commented on src/test/java/htsjdk/variant/variantcontext/VariantContextUnitTest.java:
```diff
@@ -1478,4 +1477,71 @@ public void testExtractStructuralVariationsData(final File vcfFile) {
+        Assert.assertEquals(context.getAttributeAsStringList("IntegerList", "empty"), Arrays.asList("0", "1", "2", "3"));
+        Assert.assertEquals(context.getAttributeAsStringList("DoubleList", "empty"), Arrays.asList("1.8", "1.6", "2.1"));
+        Assert.assertEquals(context.getAttributeAsStringList("StringList", "empty"), Arrays.asList("1", "2"));
+        Assert.assertEquals(context.getAttributeAsStringList("NotNumeric", "empty"), Arrays.asList("A", "B"));
```
please test an empty list and a missing list (here and in the other tests too!)

magicDGS commented on src/main/java/htsjdk/variant/variantcontext/CommonInfo.java:
```diff
@@ -242,19 +244,64 @@ public Object getAttribute(String key, Object defaultValue) {
+
+    public List<Double> getAttributeAsDoubleList(String key, Double defaultValue) {
+        return getAttributeAsList(key, x -> {
+                    if (x == null || x == VCFConstants.MISSING_VALUE_v4) {
```
Done

magicDGS commented:
Addressed the comments. Back to you @yfarjoun!

yfarjoun commented on src/test/java/htsjdk/variant/variantcontext/VariantContextUnitTest.java:
```diff
@@ -1478,4 +1477,76 @@ public void testExtractStructuralVariationsData(final File vcfFile) {
+        Assert.assertEquals(context.getAttributeAsStringList("StringList", "empty"), Arrays.asList("1", "2"));
+        Assert.assertEquals(context.getAttributeAsStringList("NotNumeric", "empty"), Arrays.asList("A", "B"));
+        // test the case of a missing key
+        Assert.assertTrue(context.getAttributeAsStringList("MissingList", "empy").isEmpty());
```
typo? empy->empty (not that it matters much)

yfarjoun commented:
tiny typo and 👍

magicDGS commented:
Fixed the typo! Thanks for the review!

yfarjoun commented:
thank-you for this!
